[FLINK-20356][docs] Updated Mesos documentation. (#14258)
apache/flink: docs/deployment/resource-providers/mesos.md

* This will be replaced by the TOC
{:toc}

## Getting Started

This *Getting Started* section guides you through setting up a fully functional Flink Cluster on Mesos.

### Introduction

[Apache Mesos](http://mesos.apache.org/) is another resource provider supported by 
Apache Flink. Flink utilizes the worker's provided by Mesos to run its TaskManagers.

> Apache Flink provides the script `bin/mesos-appmaster.sh` to create the Flink 

Author: Good that you mention that. I struggled to verify that firstly even though the old docs mentioned it as well. Hence, I didn't add it in my version of the docs. 
But trying to run the script from an Mesos-independent instance worked now. I adapted the docs to reflect it.

on Mesos cluster.

### Preparation

Flink on Mesos expects a Mesos cluster to be around. It also requires the Flink binaries being deployed

> ontothe the Mesos master. Additionally, Hadoop needs to be installed on the very same machine.

Author: I verified it: Using the bundled jars would be another approach to achieve it. I'm gonna add a Hadoop section similar to YARN's Hadoop sections.

Flink provides `bin/mesos-appmaster.sh` to create a Flink on Mesos cluster. It will instantiate a 

> JobManager process on the Mesos master. The Mesos workers will be utilized to run the TaskManager 

Author: I adapted the docs accordingly.

2. **Using Docker containerization**: This enables the user to provide user libraries and other 
   customizations as part of a Docker image. Docker utilization can be enabled by setting 
   [mesos.resourcemanager.tasks.container.type]({% link deployment/config.md %}#mesos-resourcemanager-tasks-container-type) 
   to `docker` and by providing the image name through [mesos.resourcemanager.tasks.container.image.name]({% link deployment/config.md %}#mesos-resourcemanager-tasks-container-image-name).

### High Availability on Mesos

You will need to run a service like Marathon or Apache Aurora which takes care of restarting the 
JobManager process in case of node or process failures. In addition, Zookeeper needs to be configured 
as described in the [High Availability section of the Flink docs]({% link deployment/ha/index.md %}).

#### Marathon

Marathon needs to be set up to launch the `bin/mesos-appmaster.sh` script. In particular, it should 
also adjust any configuration parameters for the Flink cluster.

Here is an example configuration for Marathon:
{% highlight javascript %}
{
    "id": "flink",

>     "cmd": "/opt/flink-1.11.2/bin/mesos-appmaster.sh -Dmesos.resourcemanager.framework.user=root -Dmesos.resourcemanager.tasks.taskmanager-cmd=/opt/flink-1.11.2/bin/mesos-taskmanager.sh -Dmesos.master=master:5050 -Djobmanager.memory.process.size=1472m -Dtaskmanager.memory.process.size=3500m -Dtaskmanager.numberOfTaskSlots=2 -Dparallelism.default=2",

Reviewer: Do we have to specify all the memory configurations or can we use the default values?

Reviewer: Where comes `master:5050` from?

    "cpus": 2,
    "mem": 1024,
    "disk": 0,
    "instances": 1,
    "env": {
        "MESOS_NATIVE_JAVA_LIBRARY": "/usr/lib/libmesos.so",

>         "HADOOP_CLASSPATH": "/opt/hadoop-2.10.1/etc/hadoop:/opt/hadoop-2.10.1/share/hadoop/common/lib/*:/opt/hadoop-2.10.1/share/hadoop/common/*:/opt/hadoop-2.10.1/share/hadoop/hdfs:/opt/hadoop-2.10.1/share/hadoop/hdfs/lib/*:/opt/hadoop-2.10.1/share/hadoop/hdfs/*:/opt/hadoop-2.10.1/share/hadoop/yarn:/opt/hadoop-2.10.1/share/hadoop/yarn/lib/*:/opt/hadoop-2.10.1/share/hadoop/yarn/*:/opt/hadoop-2.10.1/share/hadoop/mapreduce/lib/*:/opt/hadoop-2.10.1/share/hadoop/mapreduce/*:/contrib/capacity-scheduler/*.jar"

Reviewer: Maybe it is simpler (for the example) to use Flink's bundled Hadoop dependency. That way we might work around the problem that Mesos relies on Hadoop even if you don't use it.

    },
    "healthChecks": [
        {
            "protocol": "HTTP",
            "path": "/",
            "port": 8081,
            "gracePeriodSeconds": 300,
            "intervalSeconds": 60,
            "timeoutSeconds": 20,
            "maxConsecutiveFailures": 3
        }
    ],
    "user": "root"
}

> Flink is installed into `/opt/flink-1.11.2` for this example having `root` as the owner of the Flink 

Reviewer: I think we can refer to the current version via
```suggestion
Flink is installed into `/opt/flink-{{site.version}}` for this example having `root` as the owner of the Flink 
```

The same applies to the Marathon configuration provided above.

* This will be replaced by the TOC
{:toc}

## Getting Started

This *Getting Started* section guides you through setting up a fully functional Flink Cluster on Mesos.

### Introduction

[Apache Mesos](http://mesos.apache.org/) is another resource provider supported by 
Apache Flink. Flink utilizes the worker's provided by Mesos to run its TaskManagers.
Apache Flink provides the script `bin/mesos-appmaster.sh` to initiate the Flink 
on Mesos cluster.

### Preparation

Flink on Mesos expects a Mesos cluster to be around. It also requires the Flink binaries being 

> deployed. Additionally, Hadoop needs to be installed on the very same machine.

Reviewer: Is it also good enough if we add the Hadoop dependencies to Flink? Then the users don't have to install Hadoop. Maybe we could say that Hadoop needs to be installed or you have to provide the Hadoop dependencies.

Author: Yes, using the bundled hadoop uber jar would work as well. I added the  `#supported-hadoop-versions` to the `Flink on Mesos Reference` section at the end of this page. I didn't mention it here since we claim the `HADOOP_CLASSPATH` to be the preferred approach.

Reviewer: Fair enough.

Marathon needs to be set up to launch the `bin/mesos-appmaster.sh` script. In particular, it should 
also adjust any configuration parameters for the Flink cluster.

Here is an example configuration for Marathon:
{% highlight javascript %}
{
  "id": "flink",
  "cmd": "/opt/flink-{{ site.version }}/bin/mesos-appmaster.sh -Djobmanager.rpc.address=$HOST -Dmesos.resourcemanager.framework.user=root -Dmesos.master=<mesos-master>:5050 -Dparallelism.default=2",
  "user": "root",
  "cpus": 2,
  "mem": 2048,
  "instances": 1,
  "env": {
    "MESOS_NATIVE_JAVA_LIBRARY": "/usr/lib/libmesos.so"
  },
  "healthChecks": [
    {
      "protocol": "HTTP",
      "path": "/",
      "port": 8081,
      "gracePeriodSeconds": 300,
      "intervalSeconds": 60,
      "timeoutSeconds": 20,
      "maxConsecutiveFailures": 3
    }
  ]
}
{% endhighlight %}

Flink is installed into `/opt/flink-{{ site.version }}` having `root` as the owner of the Flink 
directory (notice that the user is used twice: once as a Marathon and another time as a Mesos 
parameter) for the example configuration above to work.

`<mesos-master>` needs to be set to the hostname or IP of Mesos' master node. `$HOST` is a Marathon 

> environment variable referring to the hostname of the machine the script is executed on.

Reviewer: So $HOST is solving the problem? Nice :-)

Author: Yes, the Marathon environment variables are going to be resolved before executing the command.

* This will be replaced by the TOC
{:toc}

> ## Getting Started

Reviewer 2: In my mind, the `Getting Started` section should be an opinionated (make choices for the user) and short (ideally doable in 15 minutes) path to having a running Flink cluster.

In particular, I would shorten the Introduction by removing the detailed description of the components, and move them to the reference.
Secondly, I would only describe how to set up on dcos, and again move the manual setup to the reference. This way, it'll just be a few commands and little text to read to success.

If you don't have a running DC/OS cluster, please follow the
[instructions on how to install DC/OS on the official website](https://dcos.io/install/).

Once you have a DC/OS cluster, you may install Flink through the DC/OS Universe. Just 
search for Flink in the search prompt. Alternatively, you can use the DC/OS CLI:

{% highlight bash %}
dcos package install flink
{% endhighlight %}

Further information can be found in the

> [DC/OS examples documentation](https://github.com/dcos/examples/tree/master/flink).

Reviewer: The documentation here installs Flink 1.3.1 (https://github.com/dcos/examples/tree/master/flink/1.10). Do we really want to promote this? If there is no newer DC/OS package, then I'd suggest to go for the option w/o DC/OS and the latest Flink version.

Author: I removed the DC/OS section and focused on deploying Flink without additional tools.

hosts executing the respective components (assuming SSH access being available for these nodes).  

Next, you have to create `${MESOS_HOME}/etc/mesos/mesos-master-env.sh` (a template file is 
provided in the same directory). In `mesos-master-env.sh`, you have to set `${MESOS_work_dir}` 
pointing to a directory used by Mesos to store data. Additionally, it is recommended to specify 
`${MESOS_log_dir}`:

{% highlight bash %}
export MESOS_work_dir=/path/to/mesos/working/directory
export MESOS_log_dir=/path/to/mesos/logging/directory
{% endhighlight %}

The Mesos agents are configured by creating `${MESOS_HOME}/etc/mesos/mesos-agent-env.sh` (a 
template file is provided in the same directory). `${MESOS_master}` needs to be configured as part 
of `mesos-agent-env.sh`. Additionally, `${MESOS_work_dir}` and `${MESOS_log_dir}` should be set to 
the corresponding folders:

{% highlight bash %}
export MESOS_master=MASTER_HOSTNAME:MASTER_PORT
export MESOS_work_dir=/path/to/mesos/working/directory
export MESOS_log_dir=/path/to/mesos/logging/directory
{% endhighlight %}

In order to start your Mesos cluster, use the deployment script `${MESOS_HOME}/sbin/mesos-start-cluster.sh`.

> Stopping the Mesos cluster can be achieved through `${MESOS_HOME}/sbin/mesos-stop-cluster.sh`.

Reviewer: Is this information maybe contained in the getting started guide for Mesos itself?

Author: You're right. I removed the information on how to install Mesos.

More information about the deployment scripts can be found [here](http://mesos.apache.org/documentation/latest/deploy-scripts/).

#### Pre-installing Flink vs Docker/Mesos containers

You may install Flink on all of your Mesos Master and Agent nodes. You can also pull the binaries 
from the Flink website during deployment and apply your custom configuration before launching the 
application master. A more convenient and easier-to-maintain approach is to use Docker containers 

> to manage the Flink binaries and configuration.

Reviewer: I'd say to go with the docker image approach.

   new TaskManager.

For Flink on Mesos to work properly, you have to export `${MESOS_NATIVE_JAVA_LIBRARY}` pointing to 
Mesos' native Java library. The location of this library differs depending on the underlying system:
- Linux: `export MESOS_NATIVE_JAVA_LIBRARY=${MESOS_HOME}/lib/libmesos.so`
- MacOS: `export MESOS_NATIVE_JAVA_LIBRARY=${MESOS_HOME}/lib/libmesos.dylib`

In order to run the `mesos-appmaster.sh` script you have to define `mesos.master` in the 
`flink-conf.yaml` or pass it via `-Dmesos.master=...` to the Java process. It is possible to 
completely parameterize a Mesos application through Java properties passed to the Mesos application 
master:
{% highlight bash %}
${FLINK_HOME}/bin/mesos-appmaster.sh \
    -Dmesos.master=localhost:5050 \
    -Dparallelism.default=10 \
    -Dtaskmanager.numberOfTaskSlots=2
{% endhighlight %}
This will create a JobManager on the machine that is used for executing the script.
In contrast to that, the TaskManagers will run as Mesos tasks in the Mesos cluster.

## Deployment Modes Supported by Flink on Mesos

For production use, we recommend deploying Flink Applications in the 

> [Per-Job]({% link deployment/index.md %}#per-job-mode), as this mode provides a better isolation 

Reviewer 2: ```suggestion
[Per-Job]({% link deployment/index.md %}#per-job-mode) mode, as this mode provides a better isolation 
```

for each job.

### Application Mode

Flink on Mesos does not support Application Mode.

### Per-Job Cluster Mode

> A job which is executed in Per-Job Cluster Mode spins up a dedicated Flink cluster that only being 

Reviewer 2: ```suggestion
A job which is executed in Per-Job Cluster Mode spins up a dedicated Flink cluster that is only being 
```

}
{% endhighlight %}

<span class="label label-info">Note</span> Make sure that all the classpath's of all Mesos 
processes contain the reference to the user code jar. There are two ways to achieve this:
1. One way is to put them in the `${FLINK_HOME}/lib/` directory, which will result in the user 
   code jar being loaded by the system classloader.

> 1. Another option is to create a `${FLINK_HOME}/usrlib/` directory and to put the user code jar 

Reviewer: ```suggestion
2. Another option is to create a `${FLINK_HOME}/usrlib/` directory and to put the user code jar 
```

   in that directory. The user code jar will be loaded by the user code classloader after launching 

>    a Per-Job Cluster via `${FLINK_HOME}/bin/mesos-appmaster-job.sh ...`.

Reviewer: Have you tried it out whether it works?

deploying Flink in Session Mode.

## Flink on Mesos Reference

### Configuring Flink on Mesos

For a list of Mesos-specific configuration parameters, refer to the 
[Mesos section]({% link deployment/config.md %}#mesos) of the configuration documentation.

### High Availability on Mesos

You will need to run a service like Marathon or Apache Aurora which takes care of restarting the 
JobManager process in case of node or process failures. In addition, Zookeeper needs to be 
configured as described in the 
[High Availability section of the Flink docs]({% link deployment/ha/index.md %}).

#### Marathon

Marathon needs to be set up to launch the `bin/mesos-appmaster.sh` script. In particular, it should 
also adjust any configuration parameters for the Flink cluster.

Here is an example configuration for Marathon:

{% highlight javascript %}
    {
        "id": "flink",
        "cmd": "$FLINK_HOME/bin/mesos-appmaster.sh -Djobmanager.memory.process.size=1472m -Djobmanager.rpc.port=6123 -Drest.port=8081 -Dtaskmanager.memory.process.size=1024m -Dtaskmanager.numberOfTaskSlots=2 -Dparallelism.default=2 -Dmesos.resourcemanager.tasks.cpus=1",
        "cpus": 1.0,
        "mem": 1024

>     }

Reviewer: I think this yaml is not correct. We allocate `mem: 1024` but set `-Djobmanager.memory.process.size=1472m`. Maybe we can put the simplest definition possible here (e.g. w/o specifying the resources and the ports, for example) or we completely remove it and refer to the documentation of Marathon.
